host.go

```diff
@@ -108,40 +128,117 @@ func (h *gpbftRunner) ValidateMessage(msg *gpbft.GMessage) (gpbft.ValidatedMessa
 // The chain should be a suffix of the last chain notified to the host via
 // ReceiveDecision (or known to be final via some other channel).
 func (h *gpbftHost) GetProposalForInstance(instance uint64) (*gpbft.SupplementalData, gpbft.ECChain, error) {
-	// TODO: this is just a complete fake
+	var baseTsk gpbft.TipSetKey
+	if instance == 0 {
+		ts, err := h.client.ec.GetTipsetByEpoch(h.runningCtx,
+			h.manifest.BootstrapEpoch-finality)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting boostrap base: %w", err)
+		}
+		baseTsk = ts.Key()
+	} else {
+		cert, err := h.client.certstore.Get(h.runningCtx, instance-1)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting cert for previous instance(%d): %w", instance-1, err)
+		}
+		baseTsk = cert.ECChain.Head().Key
+	}
 
-	pt, _, err := h.GetCommitteeForInstance(0)
+	baseTs, err := h.client.ec.GetTipset(h.runningCtx, baseTsk)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("getting power table: %w", err)
+		return nil, nil, xerrors.Errorf("getting base TS: %w", err)
+	}
+	headTs, err := h.client.ec.GetHead(h.runningCtx)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("getting head TS: %w", err)
+	}
+
+	collectedChain, err := h.collectChain(baseTs, headTs)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("collecting chain: %w", err)
+	}
+
+	base := gpbft.TipSet{
+		Epoch: baseTs.Epoch(),
+		Key:   baseTs.Key(),
+	}
+	pte, err := h.client.ec.GetPowerTable(h.runningCtx, baseTs.Key())
+	if err != nil {
+		return nil, nil, xerrors.Errorf("getting power table for base: %w", err)
+	}
+	base.PowerTable, err = certs.MakePowerTableCID(pte)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("computing powertable CID for base: %w", err)
 	}
-	ptCid, err := certs.MakePowerTableCID(pt.Entries)
+
+	suffix := make([]gpbft.TipSet, min(gpbft.CHAIN_MAX_LEN-1, len(collectedChain))) // -1 because of base
+	for i := 0; i < len(suffix) && i < len(collectedChain); i++ {
```

Comment:
```suggestion
	for i := range suffix {
```

Reply:
thanks, it is a remnant from when I didn't `min` the suffix length